connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/writer/BigtableFlushableWriter.java

```diff
@@ -33,26 +51,82 @@
+    private Counter numOutEntryFailuresCounter;
+    private Counter numBatchFailuresCounter;
+    private Histogram numEntriesPerFlush;
+
+    private static final Logger logger = LoggerFactory.getLogger(BigtableFlushableWriter.class);
+
+    private static final Integer BATCHER_CLOSE_TIMEOUT_SECONDS = 60;
 
     public BigtableFlushableWriter(
-            BigtableDataClient client, Sink.InitContext sinkInitContext, String table) {
+            BigtableDataClient client, WriterInitContext sinkInitContext, String table) {
         this.client = client;
+        this.table = table;
         this.batcher = client.newBulkMutationBatcher(TableId.of(table));
-        this.context = sinkInitContext;
+
+        // Instantiate Metrics
+        this.numRecordsOutCounter =
+                sinkInitContext.metricGroup().getIOMetricGroup().getNumRecordsOutCounter();
+        this.numBytesOutCounter =
+                sinkInitContext.metricGroup().getIOMetricGroup().getNumBytesOutCounter();
+        this.numOutEntryFailuresCounter =
+                sinkInitContext.metricGroup().counter("numOutEntryFailuresCounter");
+        this.numBatchFailuresCounter =
+                sinkInitContext.metricGroup().counter("numBatchFailuresCounter");
+        this.numEntriesPerFlush =
+                sinkInitContext
+                        .metricGroup()
+                        .histogram("numEntriesPerFlush", new DescriptiveStatisticsHistogram(100));
     }
 
     /** Adds RowMuationEntry to Batcher. */
     public void collect(RowMutationEntry entry) throws InterruptedException {
-        batcher.add(entry);
+        ApiFuture<Void> future = batcher.add(entry);
+        batchFutures.add(future);
+        totalRecordsBuffer++;
+        totalBytesBuffer += getEntryBytesSize(entry);
     }
 
     /** Sends mutations to Bigtable. */
     public void flush() throws InterruptedException {
-        batcher.flush();
+        try {
+            batcher.close(Duration.ofSeconds(BATCHER_CLOSE_TIMEOUT_SECONDS));
+            // Update metrics
+            this.numEntriesPerFlush.update(totalRecordsBuffer);
+            this.numRecordsOutCounter.inc(totalRecordsBuffer);
+            this.numBytesOutCounter.inc(totalBytesBuffer);
+            // Recreate client and clear metrics
+            batcher = client.newBulkMutationBatcher(TableId.of(this.table));
+            batchFutures.clear();
+            totalRecordsBuffer = 0L;
+            totalBytesBuffer = 0L;
+        } catch (BatchingException batchingException) {
+            this.numBatchFailuresCounter.inc();
+            for (ApiFuture<Void> future : batchFutures) {
+                try {
+                    future.get();
+                } catch (CancellationException
+                        | ExecutionException
+                        | InterruptedException entryException) {
+                    this.numOutEntryFailuresCounter.inc();
+                    logger.error(entryException.getMessage());
```

Comment:
I am a bit unsure if we should log this.  On one hand it's nice to see all errors we get in a batch, on the other hand they may be too many.

I am considering adding a MAX_LOGS_PER_BATCH

Reply:
i'd imagine if a bunch of things are failing in a batch, they will probably be the same error. we can sample for logging (ie, only log one error per batch)

Reply:
The batching error includes one, so I will remove the log